dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/internal/handlers/AndroidSpanLogsHandler.kt

```diff
@@ -15,69 +16,83 @@ internal class AndroidSpanLogsHandler(
         )
     }
 
-    override fun log(map: Map<String, *>, span: DDSpan) {
-        extractError(map, span)
-        logFields(map)
+    override fun log(fields: Map<String, *>, span: DDSpan) {
+        val mutableMap = fields.toMutableMap()
+        extractError(mutableMap, span)
+        logFields(span, mutableMap)
+    }
+
+    override fun log(timestampMicroseconds: Long, fields: Map<String, *>, span: DDSpan) {
+        val mutableMap = fields.toMutableMap()
+        extractError(mutableMap, span)
+        logFields(span, mutableMap, timestampMicroseconds)
     }
 
     // endregion
 
     // region Internal
 
-    private fun toMilliseconds(timestampMicroseconds: Long): Long {
-        return TimeUnit.MILLISECONDS.toSeconds(timestampMicroseconds)
+    private fun toMilliseconds(timestampMicroseconds: Long?): Long? {
+        return timestampMicroseconds?.let { TimeUnit.MILLISECONDS.toSeconds(it) }
     }
 
-    private fun logFields(fields: Map<String, *>, timestamp: Long? = null) {
-        if (timestamp != null) {
-            logger.internalLog(
-                Log.VERBOSE,
-                TRACE_LOG_MESSAGE,
-                null,
-                localAttributes = fields,
-                timestamp = timestamp
-            )
-        } else {
-            logger.v(TRACE_LOG_MESSAGE, attributes = fields)
-        }
+    private fun logFields(
+        span: DDSpan,
+        fields: MutableMap<String, Any?>,
+        timestampMicroseconds: Long? = null
+    ) {
+        val message = fields.remove(Fields.MESSAGE)?.toString() ?: DEFAULT_EVENT_MESSAGE
+        fields[LogSerializer.TAG_TRACE_ID] = span.traceId.toString()
+        fields[LogSerializer.TAG_SPAN_ID] = span.spanId.toString()
+        logger.internalLog(
+            Log.VERBOSE,
+            message,
+            null,
+            fields,
+            toMilliseconds(timestampMicroseconds)
+        )
     }
 
     private fun extractError(
-        map: Map<String, *>,
+        map: MutableMap<String, *>,
         span: DDSpan
     ) {
-        val throwable = map[ERROR_OBJECT] as? Throwable
-        if (throwable != null) {
-            span.setErrorMeta(throwable)
-        } else {
-            (map[MESSAGE] as? String)?.let {
-                span.setTag(DDTags.ERROR_MSG, it)
-            }
+        val throwable = map.remove(Fields.ERROR_OBJECT) as? Throwable
+        val kind = map.remove(Fields.ERROR_KIND)
+        val errorType = kind?.toString() ?: throwable?.javaClass?.name
+
+        if (errorType != null) {
+            val stackField = map.remove(Fields.STACK)
+            val msgField = map.remove(Fields.MESSAGE)
+            val stack = stackField?.toString() ?: throwable?.loggableStackTrace()
+            val message = msgField?.toString() ?: throwable?.message
+
+            span.setError(true)
+            span.setTag(DDTags.ERROR_TYPE, errorType)
+            span.setTag(DDTags.ERROR_MSG, message)
+            span.setTag(DDTags.ERROR_STACK, stack)
```

Comment:
According to OT spec, those are [Log fields](https://github.com/opentracing/specification/blob/master/semantic_conventions.md#log-fields-table), not [Span tags](https://github.com/opentracing/specification/blob/master/semantic_conventions.md#span-tags-table), no 🤔? I can imagine that several errors might occur for single span, and here seems that we do overwrite them, no?

Reply:
Usually an error happens in a try/catch block so the way I see it it will only be one.

Reply:
In the OT specs, logs are attached to a span. What we do in this ticket is extract error fields and attach them to the span directly (as Datadog attributes)